crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -128,37 +129,95 @@ impl<S: States> NextState<S> {
     }
 }
 
+/// The previous state of [`State<S>`].
+///
+/// This is used internally by Bevy and generally not meant to be used directly.
+/// Precise semantics of data in this resource are not specified.
+#[derive(Resource, Default, Debug)]
+pub struct PrevState<S: States>(pub Option<S>);
```

Comment:
We can probably unset this after calling `OnTransition`

Reply:
Aren't you as paranoid about performance of running extra system per state 60 times per second as I am?

Probably it is nothing, but this is the only argument I have against running it as a separate system after `on_enter`. From design point of view, this would be best.